webapp/src/serial.tsx

```diff
@@ -75,60 +78,76 @@ export class Editor extends srceditor.Editor {
     }
 
     setSim(b: boolean) {
-        this.isSim = b
-        this.clear()
+        if (this.isSim != b) {
+            this.isSim = b
+            this.clear()
+        }
     }
 
     constructor(public parent: pxt.editor.IProjectView) {
         super(parent)
-        window.addEventListener("message", this.processMessage.bind(this), false)
+        window.addEventListener("message", this.processEvent.bind(this), false)
         const serialTheme = pxt.appTarget.serial && pxt.appTarget.serial.editorTheme
         this.lineColors = (serialTheme && serialTheme.lineColors) || this.lineColors
     }
 
-    processMessage(ev: MessageEvent) {
+    saveMessageForLater(ev: MessageEvent) {
+        this.savedMessageQueue.push(ev);
+        if (this.savedMessageQueue.length > this.maxSavedMessages) {
+            this.savedMessageQueue.shift();
+        }
+    }
+
+    processQueuedMessages() {
+        this.savedMessageQueue.forEach(m => this.processMessage(m));
+        this.savedMessageQueue = [];
+    }
+
+    processEvent(ev: MessageEvent) {
         let msg = ev.data
-        if (!this.active || msg.type !== "serial") return;
+        if (msg.type !== "serial") return;
+
+        if (!this.active) {
+            this.saveMessageForLater(msg);
+            return;
+        }
 
+        this.processMessage(msg);
+    }
+
+    processMessage(msg: any) {
         const smsg = msg as pxsim.SimulatorSerialMessage
         const sim = !!smsg.sim
         if (sim != this.isSim) return;
 
         const data = smsg.data || ""
-        const source = smsg.id || "?"
-
-        if (!this.sourceMap[source]) {
-            let sourceIdx = Object.keys(this.sourceMap).length + 1
-            this.sourceMap[source] = lf("source") + sourceIdx.toString()
-        }
-        let niceSource = this.sourceMap[source]
 
         const m = /^\s*(([^:]+):)?\s*(-?\d+(\.\d*)?)/i.exec(data);
         if (m) {
             const variable = m[2] || '';
             const nvalue = parseFloat(m[3]);
             if (!isNaN(nvalue)) {
-                this.appendGraphEntry(niceSource, variable, nvalue)
+                this.appendGraphEntry(variable, nvalue)
```

Comment:
are we not differentiating data from different sources? what happens when 2 micro:bit stream the same data?